Continuation of PR 40097 (Input components should use SetUpdatesAttributeName)

SteveSandersonMS opened the pull request:
This is a continuation of https://github.com/dotnet/aspnetcore/pull/40326. I need to make it into a new PR because the original one is on a branch I can't directly write to. For the details about what this fixes, consider reading the conversations on the earlier PRs and the original issue that led to all this. As a very brief summary:

 * `InputText` and the other `Input*` components do *not* currently use `SetUpdatesAttributeName` during rendering, so the diffing system doesn't know what text is currently in the DOM, and hence it emits updates to the value if and only if the bound .NET value changes (regardless of what's in the DOM)
 * This leads to a problem if the .NET property setter reverts an incoming change. Doing so means the .NET value is effectively unchanged, and hence the diffing system thinks there's nothing to do, so the DOM value is left alone even though it *has* changed. The UI is then out of sync with the .NET state.
 * An earlier PR tried to fix this by just adding `SetUpdatesAttributeName`, but that led to a different problem whereby the Input components lost the ability to represent invalid states (e.g., an `InputNumber` being blank even though it's bound to an `int` which cannot have no value).
 * To avoid breaking existing apps and losing the ability to represent invalid states, the newer PR keeps track of the possibly-unparseable incoming values and reflects those back out to the DOM instead of just emitting the latest-validly-parsed model state.

@hakenr, thanks again for contributing this! I know it's take a long time to get this in. Hopefully that doesn't come as a surprise because [we previously discussed aiming to do this in an early .NET 8 preview](https://github.com/dotnet/aspnetcore/pull/40326#issuecomment-1214973556), and that's where we are now.

I think the PR at https://github.com/dotnet/aspnetcore/pull/40326 is already good and almost ready to merge. However there's a bit of remaining work to rebase on `main` and a couple of tweaks I want to make the the new public API, so I'll do that directly here if that's OK.

hakenr committed:
Input components two-way binding - InputText E2E test + fix

hakenr committed:
Input components two-way binding - InputDate E2E test + fix

hakenr committed:
revert .sln change

hakenr committed:
Input components two-way binding - InputNumber E2E test + fix

hakenr committed:
Input components two-way binding - InputCheckbox E2E test + fix

hakenr committed:
Input components two-way binding - InputTextarea E2E test + fix

hakenr committed:
Input components two-way binding - InputRadio E2E test + fix

hakenr committed:
Input components two-way binding - InputSelect E2E test + fix

hakenr committed:
Input components two-way binding - nullable reference types annotations removal

hakenr committed:
InputBase.CurrentValueAsString to handle unparsable values

hakenr committed:
Input components two-way binding - E2E tests for failed parsing + preserving CurrentValueAsString logic (refactoring to new protected properties)

SteveSandersonMS committed:
Revert some unrelated changes

SteveSandersonMS commented:
As an open design question, we have to decide where to put the possibly-unparseable incoming value so that subclasses can use it in their rendering logic. I think there are three main options:

 * Most risk-averse: We leave `CurrentValueAsString` alone (so it continues only to contain the most recent valid value) and introduce a new property `LatestUserInputWhichMightNotEventBeParseable`, with some better name. Then we recommend that people subclassing `InputBase` switch to use `LatestUserInputWhichMightNotEventBeParseable` if they can.
 * Middle option: We change `CurrentValueAsString` to represent the latest user input which may or may not be parseable. But in case someone was relying on getting the last-good-value, we add a new property `MostRecentParsedValueAsString` so people can switch to that if they need it for back-compat.
 * Most pro-risk: We avoid introducing any new API at all, and instead just change `CurrentValueAsString` to mean the latest user input, whether or not it was parseable. This is technically a breaking change, but in the vast majority of cases it would improve the behavior of existing `InputBase` subclasses. In the very rare case where someone really does need the "most recently parsed value as a string", they can use `CurrentValue` and stringify it themselves.

I was previously advocating a risk-averse approach here, but on further consideration I am now siding with the pro-risk option because I don't want to have multiple different kinds of "current value" that component authors would have to understand and distinguish. It feels better just to make the existing property represent the value people should most likely actually be using. Then we have no new API and existing subclasses will just start behaving better. But we would have to announce it as a breaking change, even if for the vast majority of people no reaction would be required.

SteveSandersonMS committed:
Update API baselines

SteveSandersonMS committed:
Just change the meaning of CurrentValueAsString, even though this is technically breaking

SteveSandersonMS committed:
No need to go through BindConverter.FormatValue if the value is already a string, since it's definitely a no-op.

The point of this change is just to make the behavior more explicit.

SteveSandersonMS committed:
Fix defective behavior I just introduced (caught by E2E tests)

hakenr commented:
@SteveSandersonMS As you are touching the `CurrentValueAsString` property and thinking about the best design, let me remind you of this issue https://github.com/dotnet/aspnetcore/issues/44105 that might affect your current design decision. I think we should shape the solution of this issue in a direction that keeps in mind the need for future asynchronous handling of input's `onchange` event.

SteveSandersonMS committed:
Add E2E cases in terms of setters that mutate incoming values (as opposed to derived inputs with custom parsers)

SteveSandersonMS committed:
In InputBase, stop mutating the Value property. Respect the one-way parameter flow like all components should.

Previously this was why setters could successfully revert inputs, but it's not the right way to do it and leads to surprising behavior (like the fact that setters could successfully revert inputs even without using SetUpdatesAttributeName). We explicitly document that components should not mutate their own inputs. Now it doesn't, and behaves more predictably but still passes tests because of SetUpdatesAttributeName.

SteveSandersonMS committed:
Workaround for InputRadioGroup issue

SteveSandersonMS committed:
Clean up the E2E tests to avoid duplication and only use custom Input* subclasses for the one scenario that needs it

SteveSandersonMS commented:
> As you are touching the CurrentValueAsString property and thinking about the best design, let me remind you of this issue https://github.com/dotnet/aspnetcore/issues/44105 that might affect your current design decision

Thanks for pointing that out. Looking into it, I think we can treat that as a separate enhancement. Whether or not we do that (and I suspect we will), it doesn't stop this PR being valid in its own right. All the test cases being added here, for example, would remain valid. Since #44105 will require some new design, I think we should address that on its own. We will know whether or not any design for #44105 still works with the requirements for this PR based on whether the E2E tests added here still pass. So in summary, I don't think we need to this PR on solving #44015.

SteveSandersonMS committed:
Rename file

SteveSandersonMS committed:
Fix for breaking TypicalValidationComponent (losing selection during typing into InputDate) and another E2E test

SteveSandersonMS committed:
Clarification

javiercn commented on src/Components/Web/src/Forms/InputBase.cs:
```diff
@@ -74,7 +76,18 @@ public abstract class InputBase<TValue> : ComponentBase, IDisposable
+                // by component depth or similar.
                 Value = value;
+
                 _ = ValueChanged.InvokeAsync(Value);
```
Side note since you mentioned https://github.com/dotnet/aspnetcore/issues/44105

Do you think it's reasonable to move the logic for invoking ValueChanged as well as `EditContext.NotifyFieldChanged` to the event handler? (Essentially manually hand-coding `@bind:after` for the `input` element binding)

That was my original idea/direction for the issue, but it is breaking-ish. That said, since we are doing a somewhat breaking change here, I think it would be a good idea to tackle both things at the same time (not in this PR) if my suggestion is something we want to follow.

SteveSandersonMS replied:
That sounds like a similar suggestion to the one raised at https://github.com/dotnet/aspnetcore/pull/46434#issuecomment-1415959720

I'm definitely open to doing that but (as per [this](https://github.com/dotnet/aspnetcore/pull/46434#issuecomment-1416085924)) would rather keep it separate.

SteveSandersonMS replied:
> Essentially manually hand-coding @bind:after for the input element binding

Ideally we might even be able to just use @bind:get/@bind:set and change the Input* components to be written as `.razor` files. I started doing that but it derailed the PR and introduced too many independent changes so I think we should tackle it separately.

akorchev replied:
@SteveSandersonMS how would one implement the equivalent of SetUpdatesAttributeName when using .razor files? I just discovered that method today why investigating an [issue](https://forum.radzen.com/t/clearing-a-textbox-after-valuechange/17323/8). It turns out this is what I need but I don't know how to use SetUpdatesAttributeName in a [razor file](https://github.com/radzenhq/radzen-blazor/blob/master/Radzen.Blazor/RadzenTextBox.razor). Is there any way or should I abandon the razor file and use BuildRenderTree?

hakenr replied:
AFAIK, `<html-element @bind-Value="..." />` in Razor generates the `SetUpdatesAttributeName()` as part of generated `BuildRenderTree()` method.

akorchev replied:
Hi @hakenr! I will test of this addresses the issue.

javiercn commented:
> I was previously advocating a risk-averse approach here, but on further consideration I am now siding with the pro-risk option because I don't want to have multiple different kinds of "current value" that component authors would have to understand and distinguish.

I agree, let's do the "right" thing since we are early in the release.

javiercn reviewed (approved):
Looks great!

SteveSandersonMS commented:
Thanks again, @hakenr! I hope you don't mind, but our branch rules require PRs to be squash-committed, so I had no choice but to do that, which is why your original commits don't show as separate commit entries in the `main` branch.

hakenr commented:
Hi @SteveSandersonMS, no problem with that. My intention was to help make Blazor better, not to hunt for position in any contribution leaderboard.

(Consider updating your workflow policies and start using `Co-authored-by:` in such cases ;-). See [GitHub: Commit together with co-authors](https://github.blog/2018-01-29-commit-together-with-co-authors/).)
